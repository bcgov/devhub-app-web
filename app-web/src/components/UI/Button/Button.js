import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import rocketchatSVG from '../../../assets/images/rocketchat_logo.svg'

const StyledButton = styled.button`
  color: #fff;
  user-select: none;
  border: 1px solid transparent;
  padding: .375rem .75rem;
  font-size: 1em;
  line-height: 1.5;
  border-radius: .25rem;
  cursor: pointer;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  display: inline-block;
  background-color: ${({variant, theme}) => {
    if(theme.colors[variant]) {
      return theme.colors[variant];
    } else {
      return 'initial';
    }
  }}
`;

const variants = {
  primary: 'primary',
  secondary: 'secondary',
  link: 'link',
}

export const Button = ({ variant, children, clicked, ...rest }) => {
  if(variant === variants.link) {
    return <StyledButton 
      onClick={clicked}  
      style={{
        backgroundColor: 'transparent',
        outline: 'none',
        color: 'inherit'
      }} 
      {...rest}>{children}</StyledButton>
  } else {
    return <StyledButton onClick={clicked} {...rest} variant={variant}>{children}</StyledButton>
  }

};

Button.propTypes = {
  children: PropTypes.node,
  clicked: PropTypes.func.isRequired,
  variant: PropTypes.oneOf(Object.keys(variants)),
};

Button.defaultProps = {
  children: '',
};


export const IconButton = styled.button`
  background-color: transparent;
  border: none;
  cursor: pointer;
  width: 35px;
  padding: 0;
  :focus, :active {
    outline: none;
  }
  :disabled {
    opacity: 0.65;
  }
`;

export const RCButton = ({...props}) => <IconButton {...props}><img src={rocketchatSVG} style={{margin: 0}}/></IconButton>

export default Button;
