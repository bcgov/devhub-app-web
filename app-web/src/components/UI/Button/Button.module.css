.Button {
    /* taken from gov bootstrap template https://github.com/bcgov/Gov-2.0-Bootstrap-Skeleton */
    display: inline-block;
    font-weight: 400;
    text-align: center;
    white-space: nowrap;
    vertical-align: middle;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    border: 1px solid transparent;
    padding: .375rem .75rem;
    font-size: 1em;
    line-height: 1.5;
    border-radius: .25rem;
    cursor: pointer;
}

.Button:focus {
    outline: none;
}
/* these classes aren't capatalized for ease of using these in dynamic logic */
.Button.primary {
    background-color: #002663;
    color: #fff;
}

.Button.secondary {
    background-color: #E3A82B;
    color: #fff;
}

.Button.link {
    background-color: transparent;
    outline: none;
    color: inherit;
}
