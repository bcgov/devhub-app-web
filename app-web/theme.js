/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING } from './src/constants/designTokens';
import { RESOURCE_TYPES, TOPICS, SEARCH_RESOURCE_TYPES, JOURNEY } from './src/constants/ui';

export const DEVHUB_PALETTE = {
  [RESOURCE_TYPES.COMPONENTS]: '#4299D2',
  [RESOURCE_TYPES.DOCUMENTATION]: '#246BD1',
  [RESOURCE_TYPES.PEOPLE]: '#E17039',
  [RESOURCE_TYPES.SELF_SERVICE_TOOLS]: '#F3BA45',
  [TOPICS]: '#444',
  [RESOURCE_TYPES.REPOSITORIES]: '#6e5494',
  [SEARCH_RESOURCE_TYPES.GITHUB_ISSUE]: '#6e5494',
  [RESOURCE_TYPES.EVENTS]: '#f44b42',
  [JOURNEY]: '#009688',
  lightgreen: '#66bb6a',
  yellow: '#FCBA19',
  lightyellow: '#F3BA45',
  grey: '#494949',
  lightgrey: '#F2F2F2',
  darkgrey: '#444',
  orange: '#E17039',
  purple: '#6E5494',
  red: '#D8292F',
  lightred: '#F44B42',
  green: '#2E8540',
  blue: '#003366',
  babyblue: '#4299D2',
  lightblue: '#246BD1',
  linkblue: '#1A5A96',
  bgBlue: '#38598A',
  white: '#ffffff',
};

export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
