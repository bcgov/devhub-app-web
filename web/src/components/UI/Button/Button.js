import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import rocketchatSVG from '../../../assets/images/rocketchat_logo.svg';
import githubPNG from '../../../assets/images/github_logo.png';
import DocumizePNG from '../../../assets/images/documize_logo.png';
import { SEARCH_SOURCES } from '../../../constants/search';

const StyledButton = styled.button`
  color: #fff;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1em;
  line-height: 1.5;
  border-radius: 0.25rem;
  cursor: pointer;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  display: inline-block;
  background-color: ${({ color, theme }) => {
    if (theme.colors[color]) {
      return theme.colors[color];
    } else {
      return 'initial';
    }
  }};
`;

const variants = {
  primary: 'primary',
  secondary: 'secondary',
  link: 'link',
};

export const TEST_IDS = {
  button: 'button',
};

export const Button = ({ variant, children, clicked, ...rest }) => {
  const colors = {
    [variants.primary]: 'blue',
    [variants.secondary]: 'yellow',
  };
  if (variant === variants.link) {
    return (
      <StyledButton
        onClick={clicked}
        style={{
          backgroundColor: 'transparent',
          outline: 'none',
          color: 'inherit',
        }}
        data-testid={TEST_IDS.button}
        {...rest}
      >
        {children}
      </StyledButton>
    );
  } else {
    return (
      <StyledButton
        onClick={clicked}
        color={colors[variant]}
        data-testid={TEST_IDS.button}
        {...rest}
      >
        {children}
      </StyledButton>
    );
  }
};

Button.propTypes = {
  children: PropTypes.node,
  clicked: PropTypes.func.isRequired,
  variant: PropTypes.oneOf(Object.keys(variants)),
};

Button.defaultProps = {
  children: '',
};

export const IconButton = styled.button`
  background-color: transparent;
  border: none;
  cursor: pointer;
  width: 35px;
  padding: 0;
  :focus,
  :active {
    outline: none;
  }
  :disabled {
    opacity: 0.65;
  }
`;

const SEARCH_SOURCES_ICON = {
  [SEARCH_SOURCES.rocketchat]: rocketchatSVG,
  [SEARCH_SOURCES.github]: githubPNG,
  [SEARCH_SOURCES.documize]: DocumizePNG,
};

export const SearchSourcesButton = props => (
  <IconButton {...props}>
    <img
      src={SEARCH_SOURCES_ICON[props.searchType]}
      style={{ margin: 0 }}
      alt={props.searchType + ' logo'}
    />
  </IconButton>
);

SearchSourcesButton.propTypes = {
  searchType: PropTypes.string.isRequired,
  style: PropTypes.object,
  title: PropTypes.string.isRequired,
};

export default Button;
