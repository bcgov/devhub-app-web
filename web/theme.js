/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING, DEVHUB_PALETTE } from './src/constants/designTokens';
export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
