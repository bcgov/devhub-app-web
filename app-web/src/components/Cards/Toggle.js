import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './Cards.module.css';
import Button from '../UI/Button/Button';

/**
 * Toggle component
 * This component has an arrow button to toggle the Cards to show all or collapse
 * @prop cardComponents containing each card
 * @prop cardLimits specify how many cards to shows as inital view
 */
class Toggle extends Component {
  state = {
    toggled: false,
  };

  toggledHandler = toggled => this.setState({ toggled });

  render() {
    const toggleIcon = this.state.toggled ? (
      <Button
        type="link"
        clicked={() => {
          this.toggledHandler(false);
        }}
      >
        Collapse
      </Button>
    ) : (
      <Button
        type="link"
        clicked={() => {
          this.toggledHandler(true);
        }}
      >
        Show All
      </Button>
    );
    const { cardComponents, cardLimits } = this.props;
    const cardShow = this.state.toggled ? cardComponents.length : cardLimits;
    const hideToggleButton =
      cardComponents.length > cardLimits ? styles.ToggleButton : styles.HideToggleButton;

    return (
      <div>
        <div className={styles.Cards}>{cardComponents.slice(0, cardShow)}</div>
        <div className={hideToggleButton}>{toggleIcon}</div>
      </div>
    );
  }
}

Toggle.propTypes = {
  cardComponents: PropTypes.array.isRequired,
  cardLimits: PropTypes.number.isRequired,
};

export default Toggle;
