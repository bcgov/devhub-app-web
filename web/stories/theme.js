import { create } from '@storybook/theming';
import { DEVHUB_PALETTE } from '../theme';

export default create({
  base: 'light',

  colorPrimary: 'hotpink',
  colorSecondary: 'deepskyblue',

  // UI
  appBg: 'white',
  appContentBg: 'white',
  appBorderColor: 'grey',
  appBorderRadius: 4,

  // Typography
  fontBase: 'sans-serif',
  fontCode: 'monospace',

  // Text colors
  textColor: DEVHUB_PALETTE.darkgrey,
  textInverseColor: 'rgba(255,255,255,0.9)',

  // Toolbar default and active colors
  barTextColor: DEVHUB_PALETTE.white,
  barSelectedColor: 'black',
  barBg: DEVHUB_PALETTE.blue,

  // Form colors
  inputBg: 'white',
  inputBorder: 'silver',
  inputTextColor: 'black',
  inputBorderRadius: 4,

  brandTitle: 'My custom storybook',
  brandImage:
    'https://u58tqm22fk-flywheel.netdna-ssl.com/wp-content/uploads/2016/06/BC-Gov-Logo-CMYK_pos.png',
});
