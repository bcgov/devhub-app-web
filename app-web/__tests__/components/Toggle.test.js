import React from 'react';
import 'jest-dom/extend-expect';
import { render, cleanup, fireEvent } from 'react-testing-library';
import { ThemeProvider } from 'emotion-theming';
import theme from '../../theme';
import { shallow } from 'enzyme';
import Toggle, { TEST_IDS } from '../../src/components/Cards/Toggle';
import Button from '../../src/components/UI/Button/Button';
describe('Toggle Component', () => {
  afterEach(cleanup)
  it('matches snapshot', () => {
    const cardLimit = 1;
    const cards = ['card1', 'card2', 'card3'];
    const {container} = render(<ThemeProvider theme={theme}><Toggle cardComponents={cards} cardLimits={cardLimit} /></ThemeProvider>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('shows collapse button when toggled on', () => {
    const cardLimit = 1;
    const cards = ['card1', 'card2', 'card3'];
    const {getByTestId} = render(<ThemeProvider theme={theme}><Toggle cardComponents={cards} cardLimits={cardLimit} /></ThemeProvider>);
    const ToggleOnButton = getByTestId(TEST_IDS.toggleOn);
    // clicks button
    expect(ToggleOnButton).toBeInTheDocument();
    fireEvent.click(ToggleOnButton);
    
    const ToggleOffButton = getByTestId(TEST_IDS.toggleOff);
    expect(ToggleOffButton).toBeInTheDocument();
    
  });


  it('hides toggle button when number of cards is less than limit', () => {
    const cardLimit = 4;
    const cards = ['card1', 'card2', 'card3'];
    const {container, rerender} = render(<ThemeProvider theme={theme}><Toggle cardComponents={cards} cardLimits={cardLimit} /></ThemeProvider>);

    expect(container.querySelectorAll('button').length).toBe(0);
    rerender(<ThemeProvider theme={theme}><Toggle cardComponents={cards} cardLimits={1} /></ThemeProvider>);
    expect(container.querySelectorAll('button').length).toBeGreaterThan(0);

  });
});
