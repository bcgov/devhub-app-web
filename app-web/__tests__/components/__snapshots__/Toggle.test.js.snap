// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Component matches snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (min-width:810px) {
  .emotion-0 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

<div
  class="emotion-0 emotion-1"
>
  card1
</div>
`;
